Re-estructure documentation (#55)
CartoDB/carto-selfhosted-helm: customizations/README.md

# Customizations

> That file explain how to configure your CARTO selfhosted to meet your needs.

Reviewer: ```suggestion
This file explains how to configure CARTO Self-Hosted to meet your needs.
```

## Production Ready

> By default, the Helm configuration with the config files provided by CARTO should work but it's not production ready.

Reviewer: ```suggestion
By default, the Helm configuration provided by CARTO works out of the box, but it's **not production ready**.
```

> In the following steps we are going to explain the steps to approximate the installation to a production ready environment.

Reviewer: ```suggestion
These are the steps to prepare the installation to a production ready environment.
```

It should be configured:
- **MANDATORY** Configure the domain to be used.
- Allow the external traffic to access the app.
- Configure DBs to be production ready. Our recomendation is to use managed DBs with backups and so on.

Configuring it is optional:
- Configure scale of the components (staticaly or with auto-scaling).
- Use your own bucket to store the data (By default, GCP CARTO buckets are used)

## Architecture diagram

<!--
We should add an arquitectural diagram to make it easier for customers to understand the parts and the relationship between them.
-->

## How to define customizations

There are two ways to configure or customize the deployment:
- [**RECOMMENDED**] Create a dedicated [yaml](https://yaml.org/) file. For example, you can create a file with the next content:
  ```yaml
  customConfigValues:
    selfHostedDomain: "my.domain.com"
  ```
  And add the following at the end of ALL the install or upgrade command:
  ```bash
  ... -f <my_customization_file>.yaml
  ```
- Use the inline set. For example, add the following at the end of ALL the install or upgrade command:
  ```bash
  ... --set customConfigValues.selfHostedDomain=my.domain.com
  ```

> ## Configure the domain of your selfhosted

Reviewer: should we add here the TLS config? I think it would make sense

Author: Mmm. I put it in the section to open to the internet. Remember that the default now it's to not be opened.

The most important step to have your CARTO self-hosted ready to be used is to configure the domain to be used.

⚠️ CARTO selfhosted is not designed to be used in the path of a URL, it needs a full domain or subdomain. ⚠️

To do this you need to [add the following customization](#how-to-define-customizations):
```yaml
customConfigValues:
  selfHostedDomain: "my.domain.com"
```

Don't forget to upgrade your chart after the change.

## Access to CARTO from outside the cluster
By default, the `router` Service of CARTO selfhosted is configured in mode `ClusterIP` so it's only usable from inside the cluster.
To access to it you can exec locally a [kubectl port-forward](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/) but this only make it accessible to your machine.

> To made it accessible from the internet (or outside the Kubernetes network) you need to properly configure the `router` Service (the easiest way) or create an Ingress.

Reviewer: I would skip the whole Ingress so far. Even say that so far the only supported way is through a `LoadBalancer`

Author: Pufff. Currently I only prepare it but I notice some shortcomings such as not being able to define a static IP (or at least I didn't see it for EKS).

At the moment I did not write anything and I am not going to do it as we said in the daily.

You decide if I remove it or leave it commented until we see what to do or leave it like this.

<!--
TODO: We need to talk about TLS and so on...
-->

We have examples for multiple cloud providers:
- [AWS EKS](service_loadBalancer/aws_eks/config.yaml)
<!--
TODO: Add the other providers
-->

### Ingress
<!--
TODO: Document Ingress
-->

### Configure HTTPS
By default, the package generate a self-signed certificate with a validity of 365 days.
Some times you need to use a valid certificate or need to totally disable it to leave the management to an external proxy.

⚠️ CARTO selfhosted only works if the final client use HTTPS protocol. ⚠️

#### Disable internal HTTPS
<!--
TODO: Document and add the ability to do it
-->

#### Configure your own certificates in CARTO

- Create a kubernetes secret with following content:
  ```bash
  kubectl create secret tls -n <namespace> <your_own_installation_name|carto>-tls-certificate --cert=path/to/cert/file --key=path/to/key/file
  ```

- [Add the following customization](#how-to-define-customizations) lines:

  ```yaml
  tlsCerts:
    autoGenerate: false
    existingSecret:
      name: "<your_own_installation_name|carto>-tls-certificate"
      keyKey: "tls.key"
      certKey: "tls.crt"
  ```

## Use external databases

> This package comes with an internal Postgres and Redis but it isn't the recommended way because WE DON'T CREATE BACKUPS.

Reviewer: ```suggestion
This package comes with an internal Postgres and Redis but it is not recommended for production. It does not have any logic for backups or any other monitoring. 
```

So we recommend to use external databases, preferible managed database by your provider, with backups, high availability, etc.

### Configure your own postgres
CARTO selfhosted require a Postges (13+) to work.
In that Postgres, CARTO stores some metadata and also the credentials of the external connections configured by the CARTO selfhosted users.

⚠️ That Postgres has nothing to do with the connections that the user configures in the CARTO workspace since it stores the metadata of the entire CARTO selfhosted. ⚠️

To proceed you need:

> - Create one secret in kubernetes with the Postgres passwords:

Reviewer 2: I think we could add anothers comments to use the external postgres and redis without create the secret manually, using [externaldb-secret.yaml](https://github.com/CartoDB/carto-selfhosted-helm/blob/main/chart/templates/externaldb-secret.yaml) and [externalredis-secret.yaml](https://github.com/CartoDB/carto-selfhosted-helm/blob/main/chart/templates/externalredis-secret.yaml)

Author: I only copy/pasted and little bit adapted from the old config. I think that this section should be rethinked/redo
